File: AiTrainerIOS/AiTrainerCore/Sources/AiTrainerCore/EbbinghausScheduler.swift
import Foundation

public enum EbbinghausScheduler {
    public static let stageNew = 0
    public static let stageMax = 10
    public static let lapseWrongThreshold = 2
    public static let defaultEase = 2.5
    public static let minEase = 1.3
    public static let maxEase = 3.0

    private static let intervalMs: [Int64] = [
        0,
        Int64(5 * 60 * 1000),
        Int64(30 * 60 * 1000),
        Int64(12 * 60 * 60 * 1000),
        Int64(24 * 60 * 60 * 1000),
        Int64(2 * 24 * 60 * 60 * 1000),
        Int64(4 * 24 * 60 * 60 * 1000),
        Int64(7 * 24 * 60 * 60 * 1000),
        Int64(15 * 24 * 60 * 60 * 1000),
        Int64(30 * 24 * 60 * 60 * 1000),
        Int64(60 * 24 * 60 * 60 * 1000),
    ]

    public static let cycleLabels = [
        "5分钟", "30分钟", "12小时", "1天", "2天", "4天", "7天", "15天", "30天", "60天",
    ]

    public static let stageLabels = ["未刷过"] + cycleLabels

    public static func defaultState(id: String) -> QuestionMemoryState {
        QuestionMemoryState(id: id)
    }

    public static func isNew(_ state: QuestionMemoryState) -> Bool {
        state.stage == stageNew && state.timesSeen == 0
    }

    public static func isLearning(_ state: QuestionMemoryState) -> Bool {
        !isNew(state) && state.reps == 0 && state.stage >= 1
    }

    public static func isDue(_ state: QuestionMemoryState, now: Int64 = currentMs()) -> Bool {
        isNew(state) || now >= state.nextReviewAt
    }

    public static func isLapse(_ state: QuestionMemoryState) -> Bool {
        state.timesWrong >= lapseWrongThreshold
    }

    public static func isEligibleForDraw(_ state: QuestionMemoryState, now: Int64 = currentMs()) -> Bool {
        isNew(state) || isDue(state, now: now) || isLapse(state)
    }

    public static func drawPriority(_ state: QuestionMemoryState, now: Int64 = currentMs()) -> Int {
        if isLapse(state) { return 0 }
        if isLearning(state) && isDue(state, now: now) { return 1 }
        if isNew(state) { return 2 }
        if isDue(state, now: now) { return 3 }
        return 99
    }

    public static func drawTieBreak(_ state: QuestionMemoryState, now: Int64 = currentMs()) -> Int64 {
        isNew(state) ? Int64.max : now - state.nextReviewAt
    }

    public static func onAnswered(
        _ state: QuestionMemoryState,
        correct: Bool,
        wasDue: Bool,
        now: Int64 = currentMs()
    ) -> QuestionMemoryState {
        if correct {
            return onCorrect(state, wasDue: wasDue, wasNew: isNew(state), now: now)
        }
        return onWrong(state, now: now)
    }

    private static func onCorrect(
        _ state: QuestionMemoryState,
        wasDue: Bool,
        wasNew: Bool,
        now: Int64
    ) -> QuestionMemoryState {
        var seen = state
        seen.timesSeen += 1
        seen.lastReviewAt = now

        if wasNew {
            let interval = scaledInterval(intervalMs[1], ease: seen.easeFactor)
            seen.stage = 1
            seen.reps = 0
            seen.intervalMs = interval
            seen.nextReviewAt = now + interval
            seen.timesWrong = 0
            return seen
        }

        if !wasDue { return state }

        let newEase = updateEase(seen.easeFactor, quality: 4)
        let newReps = seen.reps + 1
        let nextStage = min(seen.stage + 1, stageMax)
        let interval = scaledInterval(intervalMs[nextStage], ease: newEase)
        seen.stage = nextStage
        seen.reps = newReps
        seen.easeFactor = newEase
        seen.intervalMs = interval
        seen.nextReviewAt = now + interval
        seen.timesWrong = 0
        return seen
    }

    private static func onWrong(_ state: QuestionMemoryState, now: Int64) -> QuestionMemoryState {
        var seen = state
        seen.timesSeen += 1
        seen.lastReviewAt = now
        seen.timesWrong += 1
        let newEase = updateEase(seen.easeFactor, quality: 0)
        let demotedStage = isNew(state) ? 1 : max(1, seen.stage - 2)
        let interval = intervalMs[demotedStage]
        seen.stage = demotedStage
        seen.reps = 0
        seen.easeFactor = newEase
        seen.intervalMs = interval
        seen.nextReviewAt = now
        return seen
    }

    private static func updateEase(_ ease: Double, quality: Int) -> Double {
        let q = min(max(quality, 0), 5)
        let delta = 0.1 - Double(5 - q) * (0.08 + Double(5 - q) * 0.02)
        return min(max(ease + delta, minEase), maxEase)
    }

    private static func scaledInterval(_ baseMs: Int64, ease: Double) -> Int64 {
        if baseMs <= 0 { return 0 }
        let scale = ease / defaultEase
        let scaled = Int64(Double(baseMs) * scale)
        return min(max(scaled, baseMs / 2), baseMs * 2)
    }

    public static func formatNextReview(_ state: QuestionMemoryState, now: Int64 = currentMs()) -> String {
        if isNew(state) { return "未刷过" }
        if isLearning(state) { return isDue(state, now: now) ? "学习到期" : "学习中" }
        if isDue(state, now: now) { return "待复习" }
        let remain = state.nextReviewAt - now
        let minutes = remain / (60 * 1000)
        let hours = remain / (60 * 60 * 1000)
        let days = remain / (24 * 60 * 60 * 1000)
        if days >= 1 { return "\(days)天后" }
        if hours >= 1 { return "\(hours)小时后" }
        if minutes >= 1 { return "\(minutes)分钟后" }
        return "即将复习"
    }

    public static func forgetUrgency(_ state: QuestionMemoryState, now: Int64 = currentMs()) -> Float {
        if isNew(state) { return 0 }
        if isLapse(state) { return 1 }
        if isDue(state, now: now) { return 1 }
        let last = state.lastReviewAt
        let next = state.nextReviewAt
        if next <= last { return 1 }
        let progress = min(max(Double(now - last) / Double(next - last), 0), 1)
        return Float(progress * progress)
    }

    public static func isAtRisk(_ state: QuestionMemoryState, now: Int64 = currentMs()) -> Bool {
        forgetUrgency(state, now: now) >= 0.5
    }

    public static func normalizeState(_ state: QuestionMemoryState, now: Int64) -> QuestionMemoryState {
        let stage = min(max(state.stage, stageNew), stageMax)
        let ease = (minEase...maxEase).contains(state.easeFactor) ? state.easeFactor : defaultEase
        let reps: Int = {
            if state.reps > 0 { return state.reps }
            if stage <= stageNew { return 0 }
            return max(0, stage - 1)
        }()
        let interval = state.intervalMs > 0 ? state.intervalMs : (stage <= stageNew ? 0 : scaledInterval(intervalMs[stage], ease: ease))
        let nextAt: Int64 = (stage == stageNew && state.timesSeen == 0)
            ? 0
            : min(state.nextReviewAt, now + Int64(365 * 24 * 60 * 60 * 1000))
        return QuestionMemoryState(
            id: state.id,
            stage: stage,
            nextReviewAt: nextAt,
            lastReviewAt: state.lastReviewAt,
            timesSeen: max(state.timesSeen, 0),
            timesWrong: max(state.timesWrong, 0),
            reps: reps,
            easeFactor: ease,
            intervalMs: interval
        )
    }

    public static func currentMs() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }
}

File: AiTrainer/app/src/main/java/com/aitrainer/practice/data/EbbinghausScheduler.kt
package com.aitrainer.practice.data

/**
 * 严格间隔复习（SRS）：受艾宾浩斯启发的 SM-2 变体。
 *
 * - 仅「到期复习」或「首次作答」会更新排期；未到期答对不升阶。
 * - 首次答对：进入阶段 1 学习，5 分钟后到期验证。
 * - 到期答对：升阶 + 按 easeFactor 拉长间隔。
 * - 答错：回退 2 个阶段（至少阶段 1）、重置 reps、降低 ease。
 */
object EbbinghausScheduler {

    const val STAGE_NEW = 0
    const val STAGE_MAX = 10
    const val LAPSE_WRONG_THRESHOLD = 2
    const val DEFAULT_EASE = 2.5
    const val MIN_EASE = 1.3
    const val MAX_EASE = 3.0

    /** 各阶段基础间隔（毫秒）；阶段 0 无间隔。 */
    private val INTERVAL_MS = longArrayOf(
        0L,
        5 * 60 * 1000L,                   // 1：5 分钟
        30 * 60 * 1000L,                  // 2：30 分钟
        12 * 60 * 60 * 1000L,             // 3：12 小时
        24 * 60 * 60 * 1000L,             // 4：1 天
        2 * 24 * 60 * 60 * 1000L,         // 5：2 天
        4 * 24 * 60 * 60 * 1000L,         // 6：4 天
        7 * 24 * 60 * 60 * 1000L,         // 7：7 天
        15 * 24 * 60 * 60 * 1000L,        // 8：15 天
        30 * 24 * 60 * 60 * 1000L,        // 9：30 天
        60 * 24 * 60 * 60 * 1000L,        // 10：60 天
    )

    val CYCLE_LABELS = arrayOf(
        "5分钟",
        "30分钟",
        "12小时",
        "1天",
        "2天",
        "4天",
        "7天",
        "15天",
        "30天",
        "60天",
    )

    val STAGE_LABELS = arrayOf(
        "未刷过",
        *CYCLE_LABELS,
    )

    fun defaultState(id: String): QuestionMemoryState = QuestionMemoryState(id = id)

    fun isNew(state: QuestionMemoryState): Boolean = state.stage == STAGE_NEW && state.timesSeen == 0

    /** 处于首次学习验证期（已见过但未完成第一次到期复习）。 */
    fun isLearning(state: QuestionMemoryState): Boolean =
        !isNew(state) && state.reps == 0 && state.stage >= 1

    fun isDue(state: QuestionMemoryState, now: Long = System.currentTimeMillis()): Boolean =
        isNew(state) || now >= state.nextReviewAt

    fun isLapse(state: QuestionMemoryState): Boolean = state.timesWrong >= LAPSE_WRONG_THRESHOLD

    fun isEligibleForDraw(state: QuestionMemoryState, now: Long = System.currentTimeMillis()): Boolean =
        isNew(state) || isDue(state, now) || isLapse(state)

    fun drawPriority(state: QuestionMemoryState, now: Long = System.currentTimeMillis()): Int = when {
        isLapse(state) -> 0
        isLearning(state) && isDue(state, now) -> 1
        isNew(state) -> 2
        isDue(state, now) -> 3
        else -> 99
    }

    fun drawTieBreak(state: QuestionMemoryState, now: Long = System.currentTimeMillis()): Long =
        if (isNew(state)) Long.MAX_VALUE else now - state.nextReviewAt

    fun onAnswered(
        state: QuestionMemoryState,
        correct: Boolean,
        wasDue: Boolean,
        now: Long = System.currentTimeMillis(),
    ): QuestionMemoryState {
        val wasNew = isNew(state)
        return if (correct) {
            onCorrect(state, wasDue = wasDue, wasNew = wasNew, now = now)
        } else {
            onWrong(state, now = now)
        }
    }

    private fun onCorrect(
        state: QuestionMemoryState,
        wasDue: Boolean,
        wasNew: Boolean,
        now: Long,
    ): QuestionMemoryState {
        val seen = state.copy(
            timesSeen = state.timesSeen + 1,
            lastReviewAt = now,
        )

        // 首次接触且答对：进入学习期，5 分钟后验证，不升 reps
        if (wasNew) {
            val interval = scaledInterval(INTERVAL_MS[1], seen.easeFactor)
            return seen.copy(
                stage = 1,
                reps = 0,
                intervalMs = interval,
                nextReviewAt = now + interval,
                timesWrong = 0,
            )
        }

        // 严格：未到期答对 — 不升阶、不改排期（不应被抽到，防御性处理）
        if (!wasDue) return state

        val newEase = updateEase(seen.easeFactor, quality = 4)
        val newReps = seen.reps + 1
        val nextStage = minOf(seen.stage + 1, STAGE_MAX)
        val interval = scaledInterval(INTERVAL_MS[nextStage], newEase)
        return seen.copy(
            stage = nextStage,
            reps = newReps,
            easeFactor = newEase,
            intervalMs = interval,
            nextReviewAt = now + interval,
            timesWrong = 0,
        )
    }

    private fun onWrong(state: QuestionMemoryState, now: Long): QuestionMemoryState {
        val seen = state.copy(
            timesSeen = state.timesSeen + 1,
            lastReviewAt = now,
            timesWrong = state.timesWrong + 1,
        )
        val newEase = updateEase(seen.easeFactor, quality = 0)
        val demotedStage = if (isNew(state)) 1 else maxOf(1, seen.stage - 2)
        val interval = INTERVAL_MS[demotedStage]
        return seen.copy(
            stage = demotedStage,
            reps = 0,
            easeFactor = newEase,
            intervalMs = interval,
            nextReviewAt = now,
        )
    }

    /** SM-2 ease 更新；q=0 答错，q=4 到期答对。 */
    private fun updateEase(ease: Double, quality: Int): Double {
        val q = quality.coerceIn(0, 5)
        val delta = 0.1 - (5 - q) * (0.08 + (5 - q) * 0.02)
        return (ease + delta).coerceIn(MIN_EASE, MAX_EASE)
    }

    private fun scaledInterval(baseMs: Long, ease: Double): Long {
        if (baseMs <= 0L) return 0L
        val scale = ease / DEFAULT_EASE
        return (baseMs * scale).toLong().coerceIn(baseMs / 2, baseMs * 2)
    }

    fun formatNextReview(state: QuestionMemoryState, now: Long = System.currentTimeMillis()): String {
        if (isNew(state)) return "未刷过"
        if (isLearning(state)) return if (isDue(state, now)) "学习到期" else "学习中"
        if (isDue(state, now)) return "待复习"
        val remain = state.nextReviewAt - now
        val minutes = remain / (60 * 1000)
        val hours = remain / (60 * 60 * 1000)
        val days = remain / (24 * 60 * 60 * 1000)
        return when {
            days >= 1 -> "${days}天后"
            hours >= 1 -> "${hours}小时后"
            minutes >= 1 -> "${minutes}分钟后"
            else -> "即将复习"
        }
    }

    fun intervalMs(stage: Int): Long = INTERVAL_MS.getOrElse(stage.coerceIn(0, STAGE_MAX)) { 0L }

    fun forgetUrgency(state: QuestionMemoryState, now: Long = System.currentTimeMillis()): Float {
        if (isNew(state)) return 0f
        if (isLapse(state)) return 1f
        if (isDue(state, now)) return 1f
        val last = state.lastReviewAt
        val next = state.nextReviewAt
        if (next <= last) return 1f
        val progress = ((now - last).toDouble() / (next - last).toDouble()).coerceIn(0.0, 1.0)
        return (progress * progress).toFloat()
    }

    fun isAtRisk(state: QuestionMemoryState, now: Long = System.currentTimeMillis()): Boolean =
        forgetUrgency(state, now) >= 0.5f

    /** 从旧版数据补齐严格 SRS 字段。 */
    fun normalizeState(state: QuestionMemoryState, now: Long): QuestionMemoryState {
        val stage = state.stage.coerceIn(STAGE_NEW, STAGE_MAX)
        val ease = state.easeFactor.takeIf { it in MIN_EASE..MAX_EASE } ?: DEFAULT_EASE
        val reps = state.reps.coerceAtLeast(0).let { r ->
            if (r > 0) r else if (stage <= STAGE_NEW) 0 else maxOf(0, stage - 1)
        }
        val interval = state.intervalMs.takeIf { it > 0 }
            ?: if (stage <= STAGE_NEW) 0L else scaledInterval(INTERVAL_MS[stage], ease)
        val nextAt = if (stage == STAGE_NEW && state.timesSeen == 0) {
            0L
        } else {
            state.nextReviewAt.coerceAtMost(now + 365L * 24 * 60 * 60 * 1000)
        }
        return state.copy(
            stage = stage,
            easeFactor = ease,
            reps = reps,
            intervalMs = interval,
            nextReviewAt = nextAt,
            timesSeen = state.timesSeen.coerceAtLeast(0),
            timesWrong = state.timesWrong.coerceAtLeast(0),
        )
    }
}
